drivers/counter/counter_mcux_gpt.c

```diff
@@ -0,0 +1,234 @@
+/*
+ * Copyright (c) 2019, Linaro Limited.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <counter.h>
+#include <fsl_gpt.h>
+#include <logging/log.h>
+
+LOG_MODULE_REGISTER(mcux_gpt, CONFIG_COUNTER_LOG_LEVEL);
+
+struct mcux_gpt_config {
+	/* info must be first element */
+	struct counter_config_info info;
+	GPT_Type *base;
+	clock_name_t clock_source;
+};
+
+struct mcux_gpt_data {
+	counter_alarm_callback_t alarm_callback;
+	counter_top_callback_t top_callback;
+	void *alarm_user_data;
+	void *top_user_data;
+};
+
+static int mcux_gpt_start(struct device *dev)
+{
+	const struct mcux_gpt_config *config = dev->config->config_info;
+
+	GPT_StartTimer(config->base);
+
+	return 0;
+}
+
+static int mcux_gpt_stop(struct device *dev)
+{
+	const struct mcux_gpt_config *config = dev->config->config_info;
+
+	GPT_StopTimer(config->base);
+
+	return 0;
+}
+
+static u32_t mcux_gpt_read(struct device *dev)
+{
+	const struct mcux_gpt_config *config = dev->config->config_info;
+
+	return GPT_GetCurrentTimerCount(config->base);
+}
+
+static int mcux_gpt_set_alarm(struct device *dev, u8_t chan_id,
+			      const struct counter_alarm_cfg *alarm_cfg)
+{
+	const struct mcux_gpt_config *config = dev->config->config_info;
+	struct mcux_gpt_data *data = dev->driver_data;
+
+	u32_t current = mcux_gpt_read(dev);
+	u32_t ticks = alarm_cfg->ticks;
+
+	if (chan_id != 0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	if (!alarm_cfg->absolute) {
+		ticks += current;
+	}
+
+	data->alarm_callback = alarm_cfg->callback;
+	data->alarm_user_data = alarm_cfg->user_data;
+
+	GPT_SetOutputCompareValue(config->base, kGPT_OutputCompare_Channel1,
+				  ticks);
+	GPT_EnableInterrupts(config->base, kGPT_OutputCompare1InterruptEnable);
+
+	return 0;
+}
+
+static int mcux_gpt_cancel_alarm(struct device *dev, u8_t chan_id)
+{
+	const struct mcux_gpt_config *config = dev->config->config_info;
+	struct mcux_gpt_data *data = dev->driver_data;
+
+	if (chan_id != 0) {
+		LOG_ERR("Invalid channel id");
+		return -EINVAL;
+	}
+
+	GPT_DisableInterrupts(config->base, kGPT_OutputCompare1InterruptEnable);
+	data->alarm_callback = NULL;
+
+	return 0;
+}
+
+void mcux_gpt_isr(void *p)
+{
+	struct device *dev = p;
+	const struct mcux_gpt_config *config = dev->config->config_info;
+	struct mcux_gpt_data *data = dev->driver_data;
+	u32_t current = mcux_gpt_read(dev);
+	u32_t status;
+
+	status =  GPT_GetStatusFlags(config->base, kGPT_OutputCompare1Flag |
+				     kGPT_RollOverFlag);
+	GPT_ClearStatusFlags(config->base, status);
+	__DSB();
+
+	if ((status & kGPT_OutputCompare1Flag) && data->alarm_callback) {
+		GPT_DisableInterrupts(config->base,
+				      kGPT_OutputCompare1InterruptEnable);
+		data->alarm_callback(dev, 0, current, data->alarm_user_data);
```

Comment:
you should clear `data->alarm_callback` to allow setting new alarm from that callback.